### DISCUSSION TITLE
How can I enable white list login ?

### DISCUSSION THREAD

**Discussion by jpsanchezl10** [docker-mailserver/docker-mailserver#3847] [Q&A] [answered]
I only want logins from a single IP address

**Comment by polarathene** [comment, marked as answer]
This is probably something you can more easily control by using a reverse proxy like Traefik, or placing both systems within a VPN (perhaps Tailscale?).

## Postfix

For Postfix, you'll need to edit the `/etc/postfix/master.cf` config for the `submission` (port 587) and `submissions` (port 465) entries. 

### Restricting client by IP

There should be a line like `-o smtpd_client_restrictions`, you'll need to modify it to instead look like this:

```
-o { smtpd_client_restrictions=check_client_access cidr:{{172.16.42.10 OK}},reject }
```

- The `{ }` wrapping around most of the line there permits spaces, which AFAIK is required for [`check_client_access`](https://www.postfix.org/postconf.5.html#smtpd_client_restrictions) to provide the CIDR table.
- Any other client IP would not have provided `OK`, so the next rule is to `reject`.

For reference, these are what DMS has configured by default:

```
# 25:
smtpd_client_restrictions = permit_mynetworks, permit_sasl_authenticated, reject_unauth_destination
# 587 / 465:
smtpd_client_restrictions = permit_sasl_authenticated, reject
```

---

### Restrict by IP but also continue to require authentication

This below variation appears to work without needing `check_client_access`, although the space between the IP and action is still required (_preventing using this via `postconf -P` in `user-patches.sh`, which means it will be inconvenient to modify `master.cf`_):

```
-o { smtpd_client_restrictions=cidr:{{!172.16.42.10 REJECT}},permit_sasl_authenticated,reject }
```

The rule was changed with a `!` in front of the IP along with the action, which allows us to additionally add the additional check that you've provided correct credentials to submit mail for DMS to send.

### Compatibility with `postconf -P` / `postfix-master.cf` via separate CIDR file

To workaround the space issue, you can move the CIDR table to a separate file, then use our support for Postfix overrides with `postfix-master.cf`:

```
submission/inet/smtpd_client_restrictions=cidr:/etc/postfix/trusted_clients,permit_sasl_authenticated,reject
submissions/inet/smtpd_client_restrictions=cidr:/etc/postfix/trusted_clients,permit_sasl_authenticated,reject
```

`/etc/postfix/trusted_clients` (_replace with your trusted IP_):

```
!172.16.42.10 REJECT
```

---

### Compatibility with `postfix-main.cf` for CIDR table via custom parameter

With v14 of DMS, a bug will be fixed which instead would allow you to also use `postfix-main.cf` with a custom parameter instead of a separate CIDR map file like `/etc/postfix/trusted_clients`:

`postfix-main.cf`:

```
dms_trusted_clients= cidr:{{!172.16.42.10 REJECT}}, permit_sasl_authenticated, reject
```

`postfix-master.cf`:

```
submission/inet/smtpd_client_restrictions=$dms_trusted_clients
submissions/inet/smtpd_client_restrictions=$dms_trusted_clients
```

## Elsewhere

You'll also want similar config for Dovecot, and if you need to login for anything else like supervisord/rspamd web UI.

If you instead restrict access to those ports via reverse proxy instead, that'd simplify configuration for you. For example with Traefik 3.0:

```yml
services:
  dms:
    labels:
      - traefik.tcp.middlewares.trusted_clients.ipallowlist.sourcerange=172.16.42.10
      # Then for each router (authenticated ports 587/465/143/993/etc), add a label to use the middleware:
      - traefik.tcp.routers.smtp-submissions.middlewares=trusted_clients
```

NOTE: Traefik 2.x calls the same middleware `ipwhitelist`.

If you already have a reverse proxy in your setup, do note that if the original client IP is not preserved (eg via PROXY protocol), then you could not apply the Postfix advice securely, as you'd always have the reverse proxy as the client IP. Similarly if you have any other hops between the client, ensure the client IP is properly preserved 👍 

---

If all the above is too much, just use a strong password and you should be ok 👍 Fail2Ban can be enabled to block anyone else attempting to login repeatedly failing auth.

**Comment by PunkFleet** [comment]
@polarathene 
I went through the official recommended override, but it seems there is a bug here.
The first attempt to log in gets rejected, but the second login goes through with no problem.

```
2024-12-11T17:05:41.619875+00:00 mail dovecot: imap-login: Login: user=<test@example.com>, method=PLAIN, rip=27.128.28.245, lip=10.89.0.9, mpid=980, TLS, session=<I6yHmwEpqLsbgBz1>
2024-12-11T17:05:42.501298+00:00 mail dovecot: imap(test@example.com)<980><I6yHmwEpqLsbgBz1>: Disconnected: Connection closed (NAMESPACE finished 0.281 secs ago) in=58 out=568 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
2024-12-11T17:06:17.948953+00:00 mail postfix/submissions/smtpd[1039]: connect from unknown[27.128.28.245]
2024-12-11T17:06:18.254865+00:00 mail postfix/submissions/smtpd[1039]: Anonymous TLS connection established from unknown[27.128.28.245]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature ECDSA (prime256v1) server-digest SHA256
2024-12-11T17:06:20.059749+00:00 mail postfix/submissions/smtpd[1039]: disconnect from unknown[27.128.28.245] ehlo=1 auth=1 quit=1 commands=3
2024-12-11T17:06:21.874301+00:00 mail dovecot: imap-login: Login: user=<test@example.com>, method=PLAIN, rip=27.128.28.245, lip=10.89.0.9, mpid=1054, TLS, session=<NfftnQEpvrsbgBz1>
2024-12-11T17:06:22.992662+00:00 mail dovecot: imap(test@example.com)<1054><NfftnQEpvrsbgBz1>: Disconnected: Connection closed (LIST finished 0.284 secs ago) in=94 out=766 deleted=0 expunged=0 trashed=0 hdr_count=0 hdr_bytes=0 body_count=0 body_bytes=0
```
there my compose.yaml
```
...
    volumes:
      - ./docker-data/dms/mail-data/:/var/mail/:Z
      - ./docker-data/dms/mail-state/:/var/mail-state/:Z
      - ./docker-data/dms/mail-logs/:/var/log/mail/:Z
      - ./docker-data/dms/config/:/tmp/docker-mailserver/:Z
      # IP Whitelist
      - ./docker-data/dms/config/postfix/trusted_clients:/etc/postfix/trusted_clients:Z
      - /etc/localtime:/etc/localtime:ro
      - /etc/letsencrypt:/etc/letsencrypt:Z
```
and
```
[root@mail ~]# ls docker-data/dms/config
dovecot-masters.cf  dovecot-quotas.cf  opendkim  postfix  postfix-accounts.cf  postfix-master.cf  postfix-virtual.cf  user-patches.sh
[root@mail ~]# cat docker-data/dms/config/postfix-master.cf
submission/inet/smtpd_client_restrictions=cidr:/etc/postfix/trusted_clients,permit_sasl_authenticated,reject
submissions/inet/smtpd_client_restrictions=cidr:/etc/postfix/trusted_clients,permit_sasl_authenticated,reject
[root@mail ~]# cat docker-data/dms/config/postfix/trusted_clients
!164.25.121.225 REJECT
[root@mail ~]# podman exec -it mailserver cat /etc/postfix/trusted_clients
!164.25.121.225 REJECT
[root@mail ~]# 
```

I did check it many times, so I shouldn't have overlooked anything, right?

**Reply by polarathene** [reply]
For reference to any readers landing here, @PunkFleet created a separate issue to discuss: https://github.com/docker-mailserver/docker-mailserver/issues/4287
